### PR TITLE
Discrete differences between adjacent grid squares (version 2)

### PR DESCRIPTION
**Pull request by meabhnicguidhir**
New PR to finish @gavinevans work on gradients between adjacent cells.
Issue: #83 
Comments and responses on previous PR #142 

Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s)

### CODE REVIEW

**Comment by fionaRust** [comment]
Does the metadata fully describe the new data in the output cube. What does an output cube look like?
Does the name change? Do we have a convention for this?

**Comment by meabhnicguidhir** [comment]
The output is two cubes, one for each of x, y coords. The name of the input is preserved as are the input units and any cell methods/attributes on the input cube. Most coordinates are preserved, but the x,y coords will be staggered accordingly. In addition, the new cube has a new attribute: `form_of_discrete_difference` which is set to forward, and a new cell method: `discrete_difference`.

We don't really have firm conventions. Gavin and I looked for any kind of gradient etc in the cf conventions but couldn't find anything similar; the new attribute and cell_method seemed sensible to us. Maybe the name of the cube should change.

**Comment by fionaRust** [comment]
Yes, maybe the cube name should change. What about the units as well?

I understand that it is useful to add a cell method to cover this, but I think cell methods are rather tightly controlled by cf conventions. Have you had a look at this section of the cf conventions?

File: lib/improver/tests/utilities/test_DiscreteDifferenceBetweenAdjacentGridSquares.py [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+
+    def setUp(self):
+        """Set up cube."""
+        data = np.array([[1, 2, 3],
+                         [2, 4, 6],
+                         [5, 10, 15]])
+        self.cube = set_up_cube(data, "wind_speed", "m s-1")
+        self.plugin = DiscreteDifferenceBetweenAdjacentGridSquares()
+
+    def test_y_dimension(self):
```
Comment (MoseleyS) [inline lib/improver/tests/utilities/test_DiscreteDifferenceBetweenAdjacentGridSquares.py]: Some of these unit tests are effectively here three times. We only need to test the bits that are unique to each function, there is no need to check the sub-functions as they are checked elsewhere.

File: lib/improver/tests/utilities/test_DiscreteDifferenceBetweenAdjacentGridSquares.py [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+        """Test discrete differences calculated along the y dimension."""
+        input_array = np.array([[1, 2, 3],
```
Comment (MoseleyS) [inline lib/improver/tests/utilities/test_DiscreteDifferenceBetweenAdjacentGridSquares.py]: The name `input_array` is confusing. Perhaps `diff_array` or `expected_array` would be more obvious.

File: lib/improver/utilities/spatial.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+        Returns
+        -------
+        diff_along_cube : Iris.cube.Cube
+            Cube after the discrete differences have been calculated along the
+            specified axis.
+        """
+        coord_name = cube.coord(axis=coord_axis).name()
+        diff_axis = cube.coord_dims(coord_name)[0]
+        diffs_along_axis = np.diff(cube.data, axis=diff_axis)
+        diff_along_cube = self.create_discrete_difference_cube(
```
Comment (MoseleyS) [inline lib/improver/utilities/spatial.py]: `diff_along_cube` isn't a good name. `diff_cube` would be better. `diffs_along_axis` would be better as `diff_along_axis`, it matches the others better. I'm open to other names though!

File: lib/improver/utilities/spatial.py [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+        for coord in cube.derived_coords:
+            dims = cube.coord_dims(coord)
+            discrete_diff_cube.add_aux_coord(coord.copy(), dims)
+
+        # Add metadata to indicate that a discrete difference has been
+        # calculated.
+        cell_method = CellMethod("discrete_difference", coords=[coord_name])
+        discrete_diff_cube.add_cell_method(cell_method)
+        discrete_diff_cube.attributes["form_of_discrete_difference"] = (
+            "forward_difference")
```
Comment (MoseleyS) [inline lib/improver/utilities/spatial.py]: I think we need something more to differentiate this new cube. Perhaps an updated cube.name of `difference_of_X` (I think the meta-data describes what sort of difference adequately). I believe @fionaRust wants to discuss this more tomorrow, so expect a follow-up to this comment before making code changes.

File: lib/improver/tests/utilities/__init__.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
+# -*- coding: utf-8 -*-
```
Comment (MoseleyS) [inline lib/improver/tests/utilities/__init__.py]: I'm not convinced that empty files need ownership information - there is nothing to own! What does @benfitzpatrick think?

File: lib/improver/tests/utilities/test_DiscreteDifferenceBetweenAdjacentGridSquares.py [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+    cube.add_dim_coord(DimCoord(np.linspace(0, 10000, x_dimension_length),
+                                'projection_x_coordinate', units='m'), 1)
+    return cube
+
+
+class Test_create_discrete_difference_cube(IrisTest):
+
+    """Test the create_discrete_difference_cube method."""
+
+    def setUp(self):
```
Comment (meabhnicguidhir) [inline lib/improver/tests/utilities/test_DiscreteDifferenceBetweenAdjacentGridSquares.py]: `SetUp` is python unittest magic. The test framework seems to call it before each unit test, so I think `self.cube` is safe. I've tried changing `self.cube` in one test and the other tests using it still pass.

File: lib/improver/tests/utilities/test_DiscreteDifferenceBetweenAdjacentGridSquares.py (hunk 1 above)
Comment (meabhnicguidhir) [inline lib/improver/tests/utilities/test_DiscreteDifferenceBetweenAdjacentGridSquares.py]: I don't agree here. Which unit test in particular do you mean? I think the tests here are correctly just testing the code in the method they relate to.
* `Test_calculate_discrete_difference` is only testing the calculation of the difference field: it passes data into the method and checks output matches expectation for four sensible cases.
* `Test_create_discrete_difference_cube` only tests the creation of a cube from the difference array: new cube will have some new metadata and coordinates compared to the original cube. We also want to check data to ensure it hasn't been corrupted by modifying the coordinates for instance.
* `Test_process` just tests the full plugin together, and exceptions. I think this is important to check that both x, and y difference cubes are created and that nothing weird happens in process() around the other two methods.

File: lib/improver/tests/utilities/test_DiscreteDifferenceBetweenAdjacentGridSquares.py (hunk 2 above)
Comment (meabhnicguidhir) [inline lib/improver/tests/utilities/test_DiscreteDifferenceBetweenAdjacentGridSquares.py]: Changed to `diff_array`

File: lib/improver/utilities/spatial.py (hunk 3 above)
Comment (meabhnicguidhir) [inline lib/improver/utilities/spatial.py]: I think these names are a bit better; changed.

File: lib/improver/utilities/spatial.py (hunk 4 above)
Comment (meabhnicguidhir) [inline lib/improver/utilities/spatial.py]: After discussion, we've settled on these rules temporarily:
* Remove 'discrete' from all class, function, variable names as the quantity is really a difference over the grid length.
* Keep the cell_method, add the interval (1 grid length).
* Keep the input cube units. Iris won't recognise '1 grid length' as a unit.
A `TODO` has been added to the `create_difference_cube` method to remind developers this metadata may need to be changed when more formal conventions are agreed upon.

File: lib/improver/tests/utilities/__init__.py (hunk 5 above)
Comment (benfitzpatrick) [inline lib/improver/tests/utilities/__init__.py]: I am happy if they have nothing in them. They could be useful if they had a docstring to describe their package - otherwise probably not.

File: lib/improver/utilities/spatial.py [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+        coord_name = cube.coord(axis=coord_axis).name()
+        diff_axis = cube.coord_dims(coord_name)[0]
+        diff_along_axis = np.diff(cube.data, axis=diff_axis)
+        diff_cube = self.create_difference_cube(
+            cube, coord_name, diff_along_axis)
+        return diff_cube
+
+    def process(self, cube):
+        """
+        Calculate the difference along the x and y axes and return
```
Comment (fionaRust) [inline lib/improver/utilities/spatial.py]: I think it might be worth saying it uses np.diff to calculate the differences so that people can go and look up the behaviour at edges etc.

File: lib/improver/tests/utilities/test_DiscreteDifferenceBetweenAdjacentGridSquares.py [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+                             [3, 6, 9]],
+                            mask=[[1, 0, 0],
+                                  [1, 0, 0]])
+        result = self.plugin.calculate_difference(cube, "y")
+        self.assertIsInstance(result, Cube)
+        self.assertArrayEqual(result.data, expected)
+        self.assertArrayEqual(result.data.mask, expected.mask)
+
+
+class Test_process(IrisTest):
```
Comment (fionaRust) [inline lib/improver/tests/utilities/test_DiscreteDifferenceBetweenAdjacentGridSquares.py]: At some point I think we should be checking the metadata is what we expect on our output cube

Comment (fionaRust) [inline lib/improver/tests/utilities/test_DiscreteDifferenceBetweenAdjacentGridSquares.py]: Oops, ignore me. This test does exist alreay

File: lib/improver/utilities/spatial.py (hunk 7 above)
Comment (meabhnicguidhir) [inline lib/improver/utilities/spatial.py]: Added.

File: lib/improver/tests/utilities/__init__.py (hunk 5 above)
Comment (meabhnicguidhir) [inline lib/improver/tests/utilities/__init__.py]: Removed.

**Comment by fionaRust** [comment]
All tests run successfully  on the desktop.
All changes requested by Stephen completed.
Everything looks good.

File: lib/improver/utilities/spatial.py [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+            y axis.
+        diff_along_x_cube : Iris.cube.Cube
+            Cube after the differences have been calculated along the
+            x axis.
+
+        Raises
+        ------
+        InvalidCubeError if the input cube has too many dimensions.
+        """
+        if cube.ndim != 2:
```
Comment (fionaRust) [inline lib/improver/utilities/spatial.py]: Why is this necessary?

**Comment by fionaRust** [comment]
Output cube looks like this:

```
difference_of_wind_speed / (m s-1)  (projection_y_coordinate: 970; projection_x_coordinate: 1041)
     Dimension coordinates:
          projection_y_coordinate                           x                             -
          projection_x_coordinate                           -                             x
     Scalar coordinates:
          forecast_period: 43200 seconds
          forecast_reference_time: 2017-02-20 03:00:00
          height: 10.0 m
          realization: 0
          time: 2017-02-20 15:00:00
     Attributes:
          Conventions: CF-1.5
          form_of_difference: forward_difference
          grid_id: enukx_standard_v1
          institution: Met Office
          source: Met Office Unified Model
          title: Operational MOGREPS-UK Model Forecast
          um_version: 10.4
     Cell methods:
          difference: projection_x_coordinate (1 grid length)
```

File: lib/improver/utilities/spatial.py (hunk 9 above)
Comment (fionaRust) [inline lib/improver/utilities/spatial.py]: If 3D or 4D cubes can be supported without looping over slices (which I think this code can do) then I think this condition should be removed. It will then need a new unit test. Sorry!

Comment (meabhnicguidhir) [inline lib/improver/utilities/spatial.py]: This constraint has been removed and another unit test added.

**Comment by fionaRust** [comment]
The code in this pull request was initially written by @gavinevans. It was reviewed and corrected by @Katie-Howard  and @meabhnicguidhir. @MoseleyS  reviewed the changes, but I have ensured that his comments have been addressed. I have done a final review and am therefore happy to merge.
